File: typescript/gunmu/trace.ts
/**
 * 追踪所有 Nothing 的产生位置。
 * 默认关闭，零开销。
 */

interface TraceState {
  enabled: boolean;
  records: Array<{ reason: string; origin: string }>;
}

const state: TraceState = {
  enabled: false,
  records: [],
};

export function enable(): void {
  state.enabled = true;
}

export function disable(): void {
  state.enabled = false;
}

export function clear(): void {
  state.records = [];
}

export function recordNothing(reason: string, origin: string): void {
  if (!state.enabled) {
    return;
  }
  state.records.push({ reason, origin });
}

export function records(): ReadonlyArray<{ reason: string; origin: string }> {
  return [...state.records];
}

export function summary(): {
  total: number;
  byReason: Record<string, number>;
  records: ReadonlyArray<{ reason: string; origin: string }>;
} {
  const byReason: Record<string, number> = {};
  for (const r of state.records) {
    byReason[r.reason] = (byReason[r.reason] ?? 0) + 1;
  }
  return {
    total: state.records.length,
    byReason,
    records: [...state.records],
  };
}

export function report(): string {
  const s = summary();
  if (s.total === 0) {
    return "没有滚木。";
  }
  const lines: string[] = [`共 ${s.total} 个滚木：`];
  for (const [reason, count] of Object.entries(s.byReason)) {
    lines.push(`  [${count}] ${reason}`);
  }
  lines.push("");
  lines.push("产生位置：");
  for (const r of s.records) {
    lines.push(`  ${r.origin}  ->  ${r.reason}`);
  }
  return lines.join("\n");
}

File: python/gunmu/trace.py
"""追踪所有 Nothing 的产生位置。

默认关闭，零开销。
"""
from __future__ import annotations

import threading
from collections import Counter
from dataclasses import dataclass, field


@dataclass
class _TraceState:
    enabled: bool = False
    records: list[tuple[str, str]] = field(default_factory=list)
    lock: threading.Lock = field(default_factory=threading.Lock)


_state = _TraceState()


def enable() -> None:
    """开启追踪。之后所有 Nothing 都会被记录。"""
    with _state.lock:
        _state.enabled = True


def disable() -> None:
    with _state.lock:
        _state.enabled = False


def clear() -> None:
    with _state.lock:
        _state.records.clear()


def record_nothing(reason: str, origin: str) -> None:
    if not _state.enabled:
        return
    with _state.lock:
        _state.records.append((reason, origin))


def records() -> list[tuple[str, str]]:
    with _state.lock:
        return list(_state.records)


def summary() -> dict:
    """按原因统计滚木。"""
    with _state.lock:
        counter = Counter(reason for reason, _ in _state.records)
        total = len(_state.records)
        recs = list(_state.records)
    return {
        "total": total,
        "by_reason": dict(counter.most_common()),
        "records": recs,
    }


def report() -> str:
    """人类可读的报告。"""
    s = summary()
    if s["total"] == 0:
        return "没有滚木。"
    lines = [f"共 {s['total']} 个滚木："]
    for reason, count in s["by_reason"].items():
        lines.append(f"  [{count}] {reason}")
    lines.append("")
    lines.append("产生位置：")
    for reason, origin in s["records"]:
        lines.append(f"  {origin}  ->  {reason}")
    return "\n".join(lines)